File: python-template/template_fenwick_tree.py
class FenwickTree:
    """
    Fenwick Tree (Binary Indexed Tree) for prefix sums.
    Supports:
        - add(index, delta)        : Point update
        - prefix_sum(index)               : Prefix sum query [1..index]
        - range_sum(l, r)          : Sum in [l..r]
        - find_kth(k)              : Smallest idx with prefix_sum(idx) >= k  (binary lifting)
    Indexing: 1-based for internal operations.
        
    Authored By: akkisinghvi28
    
    """

    def __init__(self, n):
        """Initialise Fenwick Tree for n elements (1-based indexing)."""
        self.n = n
        self.bit = [0] * (n + 1)

    def add(self, idx, delta):
        """Add `delta` to element at position `idx`."""
        while idx <= self.n:
            self.bit[idx] += delta
            idx += idx & -idx

    def prefix_sum(self, idx):
        """Return prefix sum from 1 to `idx`."""
        res = 0
        while idx > 0:
            res += self.bit[idx]
            idx -= idx & -idx
        return res

    def range_sum(self, l, r):
        """Return range sum from `l` to `r`."""
        return self.prefix_sum(r) - self.prefix_sum(l-1)
    
    def find_kth(self, k):
        """
        Binary lifting: Find smallest index such that prefix_sum(index) >= k.
        If total sum < k, returns n+1 (invalid).
        """
        idx = 0
        bit_mask = 1 << (self.n.bit_length() - 1)
        while bit_mask:
            nxt = idx + bit_mask
            if nxt <= self.n and self.bit[nxt] < k:
                k -= self.bit[nxt]
                idx = nxt
            bit_mask >>= 1
        return idx + 1

class DoubleFenwick:
    """
    Double Fenwick Tree for range updates and range sum queries.

    Supports:
        - range_add(l, r, delta) : Add delta to all elements in [l, r]
        - prefix_sum(idx)        : Sum of elements in [1..idx]
        - range_sum(l, r)        : Sum of elements in [l..r]

    Uses two FenwickTree instances internally.
    Indexing: 1-based
    """

    def __init__(self, n):
        self.n = n
        self.B1 = FenwickTree(n)
        self.B2 = FenwickTree(n)

    def _add(self, ft, idx, delta):
        """Internal helper to safely add within bounds."""
        if idx <= self.n:
            ft.add(idx, delta)

    def range_add(self, l, r, delta):
        """
        Add `delta` to all elements in range [l, r].
        """
        # Update B1
        self._add(self.B1, l, delta)
        self._add(self.B1, r + 1, -delta)

        # Update B2
        self._add(self.B2, l, delta * (l - 1))
        self._add(self.B2, r + 1, -delta * r)

    def prefix_sum(self, idx):
        """
        Return sum of elements in range [1..idx].
        """
        return self.B1.sum(idx) * idx - self.B2.sum(idx)

    def range_sum(self, l, r):
        """
        Return sum of elements in range [l..r].
        """
        return self.prefix_sum(r) - self.prefix_sum(l - 1)


class FenwickTree2D:
    def __init__(self, n, m):
        self.n = n
        self.m = m
        self.bit = [[0] * (m + 1) for _ in range(n + 1)]

    def add(self, x, y, delta):
        i = x
        while i <= self.n:
            j = y
            while j <= self.m:
                self.bit[i][j] += delta
                j += j & -j
            i += i & -i

    def sum(self, x, y):
        res = 0
        i = x
        while i > 0:
            j = y
            while j > 0:
                res += self.bit[i][j]
                j -= j & -j
            i -= i & -i
        return res

    def range_sum(self, x1, y1, x2, y2):
        return (self.sum(x2, y2)
                - self.sum(x1 - 1, y2)
                - self.sum(x2, y1 - 1)
                + self.sum(x1 - 1, y1 - 1))

File: cpp-templates/template_fenwick_tree.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;

struct FenwickTree {
    int n;
    vector<ll> bit;

    FenwickTree(int n = 0) { init(n); }

    void init(int n_) {
        n = n_;
        bit.assign(n + 1, 0);
    }

    void add(int idx, ll delta) {
        for (; idx <= n; idx += idx & -idx) bit[idx] += delta;
    }

    ll prefix_sum(int idx) const {
        ll res = 0;
        for (; idx > 0; idx -= idx & -idx) res += bit[idx];
        return res;
    }

    ll range_sum(int l, int r) const {
        return prefix_sum(r) - prefix_sum(l - 1);
    }

    int find_kth(ll k) const {
        int idx = 0;
        int bit_mask = 1;
        while ((bit_mask << 1) <= n) bit_mask <<= 1;
        for (; bit_mask; bit_mask >>= 1) {
            int nxt = idx + bit_mask;
            if (nxt <= n && bit[nxt] < k) {
                idx = nxt;
                k -= bit[nxt];
            }
        }
        return idx + 1;
    }
};

struct DoubleFenwick {
    int n;
    FenwickTree B1, B2;

    DoubleFenwick(int n = 0) { init(n); }

    void init(int n_) {
        n = n_;
        B1.init(n);
        B2.init(n);
    }

    void add_safe(FenwickTree& ft, int idx, ll delta) {
        if (idx <= n) ft.add(idx, delta);
    }

    void range_add(int l, int r, ll delta) {
        add_safe(B1, l, delta);
        add_safe(B1, r + 1, -delta);
        add_safe(B2, l, delta * (l - 1));
        add_safe(B2, r + 1, -delta * r);
    }

    ll prefix_sum(int idx) const {
        return B1.prefix_sum(idx) * idx - B2.prefix_sum(idx);
    }

    ll range_sum(int l, int r) const {
        return prefix_sum(r) - prefix_sum(l - 1);
    }
};

struct FenwickTree2D {
    int n, m;
    vector<vector<ll>> bit;

    FenwickTree2D(int n = 0, int m = 0) { init(n, m); }

    void init(int n_, int m_) {
        n = n_;
        m = m_;
        bit.assign(n + 1, vector<ll>(m + 1, 0));
    }

    void add(int x, int y, ll delta) {
        for (int i = x; i <= n; i += i & -i) {
            for (int j = y; j <= m; j += j & -j) {
                bit[i][j] += delta;
            }
        }
    }

    ll sum(int x, int y) const {
        ll res = 0;
        for (int i = x; i > 0; i -= i & -i) {
            for (int j = y; j > 0; j -= j & -j) {
                res += bit[i][j];
            }
        }
        return res;
    }

    ll range_sum(int x1, int y1, int x2, int y2) const {
        return sum(x2, y2) - sum(x1 - 1, y2) - sum(x2, y1 - 1) + sum(x1 - 1, y1 - 1);
    }
};
